File: bindings/python/tests/test_client.py
#!/usr/bin/env python3
"""
Simple test script for CancelToken and OCIClient.

Run with: python3 tests/test_client.py
"""

import sys
import os
import tempfile

# Add parent directory to path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import crumblecracker as cc


def test_cancel_token():
    """Test cancel token creation and cancellation."""
    print("TEST: cc.CancelToken()...", end=" ")
    cc.init()
    try:
        token = cc.CancelToken()
        assert not token.is_cancelled, "New token should not be cancelled"

        token.cancel()
        assert token.is_cancelled, "Token should be cancelled after cancel()"

        token.close()
        print("PASSED")
    finally:
        cc.shutdown()


def test_cancel_token_context_manager():
    """Test cancel token as context manager."""
    print("TEST: CancelToken context manager...", end=" ")
    cc.init()
    try:
        with cc.CancelToken() as token:
            assert not token.is_cancelled
            token.cancel()
            assert token.is_cancelled
        print("PASSED")
    finally:
        cc.shutdown()


def test_oci_client_creation():
    """Test OCI client creation."""
    print("TEST: cc.OCIClient()...", end=" ")
    cc.init()
    try:
        with cc.OCIClient() as client:
            cache_dir = client.cache_dir
            assert cache_dir, "Cache dir should not be empty"
            print(f"PASSED (cache={cache_dir})")
    finally:
        cc.shutdown()


def test_oci_client_custom_cache():
    """Test OCI client with custom cache directory."""
    print("TEST: OCIClient with custom cache...", end=" ")
    cc.init()
    try:
        with tempfile.TemporaryDirectory() as tmpdir:
            cache_path = os.path.join(tmpdir, "cache")
            os.makedirs(cache_path)

            with cc.OCIClient(cache_dir=cache_path) as client:
                # The Go library may append subdirectories
                assert cache_path in client.cache_dir, \
                    f"Expected {cache_path} in {client.cache_dir}"
                print(f"PASSED (path={client.cache_dir})")
    finally:
        cc.shutdown()


def main():
    print("=== Client Tests ===\n")

    test_cancel_token()
    test_cancel_token_context_manager()
    test_oci_client_creation()
    test_oci_client_custom_cache()

    print("\n=== All client tests passed! ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: bindings/rust/tests/test_client.rs
//! OCI client tests.
//!
//! These tests do not require hypervisor access (except for VM tests).

use cc::{CancelToken, OciClient};

#[test]
fn test_cancel_token() {
    let token = CancelToken::new();

    // New token should not be cancelled
    assert!(
        !token.is_cancelled(),
        "new token should not be cancelled"
    );

    // Cancel the token
    token.cancel();

    // Now it should be cancelled
    assert!(
        token.is_cancelled(),
        "token should be cancelled after cancel()"
    );
}

#[test]
fn test_oci_client_new() {
    cc::init().expect("init should succeed");

    let client = OciClient::new().expect("OciClient::new should succeed");

    // Cache dir should not be empty
    let cache_dir = client.cache_dir();
    assert!(
        !cache_dir.is_empty(),
        "cache_dir should not be empty"
    );
    println!("Cache dir: {}", cache_dir);

    cc::shutdown();
}

#[test]
fn test_oci_client_with_cache_dir() {
    cc::init().expect("init should succeed");

    let temp_dir = std::env::temp_dir().join("cc_rust_test_cache");
    let _ = std::fs::create_dir_all(&temp_dir);

    let client = OciClient::with_cache_dir(temp_dir.to_str().unwrap())
        .expect("OciClient::with_cache_dir should succeed");

    let cache_dir = client.cache_dir();
    assert!(
        cache_dir.contains("cc_rust_test_cache"),
        "cache_dir should contain our custom directory"
    );

    cc::shutdown();
}
